### ISSUE TITLE
How to export the database with promiser?

### ISSUE THREAD

**Issue by pycan-jouza** [sqlite/sqlite-wasm#67]
Hello,

I'm trying to export the database from browser to local file system. I'm following [this guide](https://sqlite.org/wasm/doc/trunk/cookbook.md). I've initiated the database with [promiser](https://www.npmjs.com/package/@sqlite.org/sqlite-wasm#in-a-wrapped-worker-with-opfs-if-available), though. Not sure how to access `capi` to use `sqlite3_js_db_export`. 

Any help appreciated, thank you!

**Comment by tomayac** [comment]
If you persist your database to the origin private file system (OPFS), you can follow [this guide](https://web.dev/articles/origin-private-file-system#copy_a_file_from_the_origin_private_file_system_to_the_user-visible_file_system) to move the database file from the OPFS to the regular file system.

**Comment by pycan-jouza** [comment]
I'll give it a go, thank you.

**Comment by sgbeal** [comment]
> Not sure how to access `capi` to use `sqlite3_js_db_export`.

One of the promiser's features, if we can call it that, is that it specifically hides all of the sqlite API in a separate thread, so does not/cannot expose specific functions. It does, however, expose its own export method:

<https://sqlite.org/wasm/doc/trunk/api-worker1.md#method-export>

**Comment by pycan-jouza** [comment]
@sgbeal thank you, missed this in the docs. I've tried both approaches and both of them work well. The OPFS `showSaveFilePicker` is not supported in [many browsers](https://developer.mozilla.org/en-US/docs/Web/API/Window/showSaveFilePicker#browser_compatibility) yet, so I decided to go with the promiser's export.

Out of curiosity, how would you go about importing the database back into the browser (OPFS)?

**Comment by sgbeal** [comment]
> Out of curiosity, how would you go about importing the database back into the browser (OPFS)?

Importing depends on which VFS you want to eventually access the db with. We have two OPFS VFSes and they have separate import requirements. See

<https://sqlite.org/wasm/doc/tip/cookbook.md#impexp>

for details.

That said: i seriously suggest avoiding the worker1/promiser APIs for anything more than the most basic of toy apps. "Remote-controlling" sqlite in a different thread is not well-suited to any serious work, as it makes certain usage patterns (e.g. transactions) very difficult to implement properly and it imposes performance issues which may otherwise normally be avoided. The absolute best way to use sqlite3/wasm is to import it into the same thread/worker in which your "business logic" code is in, as this gives your main db logic complete access to the database API. The worker1/promiser APIs, on the other hand, provide only very limited access to the database via highly abstracted APIs.

**Comment by pycan-jouza** [comment]
> That said: i seriously suggest avoiding the worker1/promiser APIs for anything more than the most basic of toy apps. "Remote-controlling" sqlite in a different thread is not well-suited to any serious work, as it makes certain usage patterns (e.g. transactions) very difficult to implement properly and it imposes performance issues which may otherwise normally be avoided. The absolute best way to use sqlite3/wasm is to import it into the same thread/worker in which your "business logic" code is in, as this gives your main db logic complete access to the database API. The worker1/promiser APIs, on the other hand, provide only very limited access to the database via highly abstracted APIs.

Uh, this is unexpected. It goes against what is the suggested usage of the library.

`Only the worker versions allow you to use the origin private file system (OPFS) storage back-end.`

Does it mean I should either use limited worker1/promiser API with OPFS or I run the database code in the main thread without OPFS (only storing the database in memory)?

**Comment by sgbeal** [comment]
> It goes against what the suggested usage of the library is.
>
> Only the worker versions allow you to use the origin private file system (OPFS) storage back-end.

First off, that statement is not from any of the docs on the sqlite.org/wasm site. Secondly, you're confusing the term "worker1 API" with "worker versions." OPFS is only available in a JS _worker thread_, and not the main thread. Whoever wrote the above was referring to a _worker thread_, not the worker1 API (which is always written "worker1" to distinguish it from a "worker" thread).

The docs on the main site very specifically point out some of the limitations of the worker1/promiser API:

<https://sqlite.org/wasm/doc/trunk/api-worker1.md>

The sqlite JS/WASM pieces can be loaded in _any_ thread. If they are loaded in the main thread, they do _not_ have access to OPFS. If the sqlite API is loaded in a worker thread, _and_ the page meets the other requirements for OPFS, sqlite will have access to OPFS. This page covers the requirements for the two separate OPFS VFSes:

<https://sqlite.org/wasm/doc/tip/persistence.md>

Apps can load the worker1 API from the main thread, and the worker1 API can access OPFS if it's available, but the approach the worker1 API takes to "remote-controlling" the library has severe limitations.

If, however, an app loads sqlite directly into a worker thread which contains client code, the client has full access to the library and has none of the limitations of the worker1 API. That is, by far, the recommended way to use it.

**Comment by pycan-jouza** [comment]
Thank you for the detailed explanation. Not really sure what of the client code to transfer to worker thread when building React application. I'd need to use `postMessage()` to communicate with the database, which is a bit clunky. I guess I'll stick with the worker1/promiser API for now. Thanks again!

**Comment by steida** [comment]
Move to Web Worker as much code as possible. Use https://github.com/GoogleChromeLabs/comlink

https://www.youtube.com/watch?v=7Rrv9qFMWNM

**Comment by tomayac** [comment]
@sgbeal Chased the misleading statement down to [another project](https://github.com/tursodatabase/libsql-wasm-experimental/blob/main/README-upstream.md#:~:text=Only%20the%20worker%20versions%20allow%20you%20to%20use%20the%20origin%20private%20file%20system%20(OPFS)%20storage%20back%2Dend.). If the README _here_ is misleading, please advise!

**Comment by sgbeal** [comment]
> If the README here is misleading, please advise!

FWIW, i don't consider it misleading but do understand how the OP drew the wrong conclusion from "worker" vs "worker1". Maybe rephrase:

> Only the worker versions allow you to use the origin private file system (OPFS) storage back-end.

To something like:

> The origin private file system (OPFS) storage back-end is only available when sqlite is loaded in a worker thread.

Or, if you want to be more verbose:

> SQLite can be loaded in any thread, but the origin private file system (OPFS) storage back-end is only available when it is loaded in a worker thread.

?

It might (not sure) also be useful to mention that OPFS is only available in "secure contexts", linking that phrase to:

https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts
